source/extensions/transport_sockets/tls/utility.cc

```diff
@@ -71,21 +71,68 @@ Envoy::Ssl::CertificateDetailsPtr Utility::certificateDetails(X509* cert, const
   return certificate_details;
 }
 
+bool Utility::labelWildcardMatch(absl::string_view dns_label, absl::string_view pattern) {
+  if (pattern.empty()) {
+    return dns_label.empty();
+  }
+
+  // Only valid if wildcard character appear once
+  if (std::count(pattern.begin(), pattern.end(), '*') == 1) {
+    constexpr char glob = '*';
+    // Check the special case of a single * pattern, as it's common.
+    if (pattern.size() == 1 && pattern[0] == glob) {
+      return true;
+    }
+    size_t i = 0, p = 0, i_star = dns_label.size(), p_star = 0;
+    while (i < dns_label.size()) {
+      if (p < pattern.size() && dns_label[i] == pattern[p]) {
+        ++i;
+        ++p;
+      } else if (p < pattern.size() && pattern[p] == glob) {
+        i_star = i;
+        p_star = p++;
+      } else if (i_star != dns_label.size()) {
+        i = ++i_star;
+        p = p_star + 1;
+      } else {
+        return false;
+      }
+    }
+
+    while (p < pattern.size() && pattern[p] == glob) {
+      ++p;
+    }
+
+    return p == pattern.size() && i == dns_label.size();
```

Comment:
Just split `pattern` to prefix and suffix and check them? That makes code easier to read and don't need to scan the middle of the dns label.

Reply:
yes true. Made this change and the function is much simpler now. thanks.